src/Workspaces/Core/MSBuild/MSBuild/Build/ProjectBuildManager.cs

```diff
@@ -0,0 +1,296 @@
+            }
+
+            var buildParameters = new MSB.Execution.BuildParameters(_projectCollection);
+            buildParameters.Loggers = new MSB.Framework.ILogger[] { _logger };
+
+            MSB.Execution.BuildManager.DefaultBuildManager.BeginBuild(buildParameters);
+
+            _started = true;
+        }
+
+        public void Stop()
+        {
+            if (!_started)
+            {
+                throw new InvalidOperationException();
+            }
+
+            MSB.Execution.BuildManager.DefaultBuildManager.EndBuild();
+            _started = false;
+        }
+
+        public Task<MSB.Execution.ProjectInstance> BuildProjectAsync(
+            MSB.Evaluation.Project project, DiagnosticLog log, CancellationToken cancellationToken)
+        {
+            var targets = new[] { "Compile", "CoreCompile" };
+
+            return BuildProjectAsync(project, targets, log, cancellationToken);
+        }
+
+        private async Task<MSB.Execution.ProjectInstance> BuildProjectAsync(
+            MSB.Evaluation.Project project, string[] targets, DiagnosticLog log, CancellationToken cancellationToken)
+        {
+            // create a project instance to be executed by build engine.
+            // The executed project will hold the final model of the project after execution via msbuild.
+            var projectInstance = project.CreateProjectInstance();
+
+            // Verify targets
+            foreach (var target in targets)
+            {
+                if (!projectInstance.Targets.ContainsKey(target))
+                {
+                    log.Add(string.Format(WorkspaceMSBuildResources.Project_does_not_contain_0_target, target), projectInstance.FullPath);
+                    return projectInstance;
+                }
+            }
+
+            _logger.SetProjectAndLog(projectInstance.FullPath, log);
+
+            var buildRequestData = new MSB.Execution.BuildRequestData(projectInstance, targets);
+
+            var result = await BuildAsync(buildRequestData, cancellationToken).ConfigureAwait(false);
+
+            if (result.OverallResult == MSB.Execution.BuildResultCode.Failure)
+            {
+                if (result.Exception != null)
+                {
+                    log.Add(result.Exception, projectInstance.FullPath);
+                }
+            }
+
+            return projectInstance;
+        }
+
+        // this lock is static because we are using the default build manager, and there is only one per process
+        private static readonly SemaphoreSlim s_buildManagerLock = new SemaphoreSlim(initialCount: 1);
+
+        private async Task<MSB.Execution.BuildResult> BuildAsync(MSB.Execution.BuildRequestData requestData, CancellationToken cancellationToken)
+        {
+            // only allow one build to use the default build manager at a time
+            using (await s_buildManagerLock.DisposableWaitAsync(cancellationToken).ConfigureAwait(continueOnCapturedContext: false))
```

Comment:
reason this suddenly has named argument?

Reply:
Refactored from old code.

Reply:
I removed these names to be consistent.